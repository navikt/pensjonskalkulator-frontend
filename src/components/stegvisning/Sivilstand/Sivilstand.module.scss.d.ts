declare const classNames: {
  readonly ingress: "ingress";
  readonly navigation: "navigation";
  readonly selectSivilstand: "selectSivilstand";
  readonly "navds-select__container": "navds-select__container";
};
export = classNames;
