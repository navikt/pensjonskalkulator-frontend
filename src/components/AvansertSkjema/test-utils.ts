export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
