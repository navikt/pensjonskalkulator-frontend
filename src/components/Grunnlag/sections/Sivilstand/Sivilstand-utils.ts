export const formatSivilstand = (sivilstand: Sivilstand): string => {
  switch (sivilstand) {
    case 'GIFT': {
      return 'Gift'
    }
    case 'ENKE_ELLER_ENKEMANN': {
      return 'Enke / Enkemann'
    }
    case 'SKILT': {
      return 'Skilt'
    }
    case 'SEPARERT': {
      return 'Separert'
    }
    case 'REGISTRERT_PARTNER': {
      return 'Registrert partner'
    }
    case 'SEPARERT_PARTNER': {
      return 'Separert partner'
    }
    case 'SKILT_PARTNER': {
      return 'Skilt partner'
    }
    case 'GJENLEVENDE_PARTNER': {
      return 'Gjenlevende partner'
    }
    default: {
      return 'Ugift'
    }
  }
}
