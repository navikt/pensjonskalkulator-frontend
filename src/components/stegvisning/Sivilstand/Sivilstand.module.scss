@use '../../../scss/variables';

.ingress {
  margin-bottom: var(--a-spacing-6);
}

.button {
  margin: var(--a-spacing-6) var(--a-spacing-6) 0 0;
}

.selectSivilstand :global(.navds-select__container) {
  width: variables.$input-width-m;
}

.selectSivilstand select {
  width: variables.$input-width-m;
}

.selectSivilstand label {
  width: 100%;
}
