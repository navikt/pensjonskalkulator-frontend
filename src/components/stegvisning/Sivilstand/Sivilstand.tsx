import React, { FormEvent, useMemo, useState } from 'react'
import { FormattedMessage, useIntl } from 'react-intl'

import {
  BodyLong,
  Button,
  Heading,
  Radio,
  RadioGroup,
  Select,
  VStack,
} from '@navikt/ds-react'

import { STEGVISNING_FORM_NAMES } from '../utils'
import { Card } from '@/components/common/Card'
import { paths } from '@/router/constants'
import { formatInntekt } from '@/utils/inntekt'
import { logger, wrapLogger } from '@/utils/logging'
import {
  sivilstandOptions,
  formatSivilstand,
  getSivilstandTekst,
  checkUkjentSivilstand,
} from '@/utils/sivilstand'

import styles from './Sivilstand.module.scss'

interface Props {
  sivilstandFolkeregister: UtvidetSivilstand
  grunnbelop?: number
  sivilstand: UtvidetSivilstand
  epsHarInntektOver2G: boolean | null
  epsHarPensjon: boolean | null
  onCancel?: () => void
  onPrevious: () => void
  onNext: (sivilstandData: {
    sivilstand: UtvidetSivilstand
    epsHarPensjon: boolean | null
    epsHarInntektOver2G: boolean | null
  }) => void
}

const convertBooleanToBooleanRadio = (
  input: boolean | null
): BooleanRadio | null => (input !== null ? (input ? 'ja' : 'nei') : null)

const convertBooleanRadioToBoolean = (
  input: BooleanRadio | null
): boolean | null => {
  if (input === null) {
    return null
  }
  return input === 'ja' ? true : false
}

export function Sivilstand({
  sivilstandFolkeregister,
  grunnbelop,
  sivilstand,
  epsHarInntektOver2G,
  epsHarPensjon,
  onCancel,
  onPrevious,
  onNext,
}: Props) {
  const intl = useIntl()
  const [validationError, setValidationError] = useState<{
    sivilstand?: string
    epsHarPensjon?: string
    epsHarInntektOver2G?: string
  }>({
    sivilstand: undefined,
    epsHarPensjon: undefined,
    epsHarInntektOver2G: undefined,
  })

  // TODO: Mappe UOPPGITT og UNKNOWN til UGIFT?
  const [sivilstandInput, setSivilstandInput] = useState(sivilstand)
  const [epsHarPensjonInput, setEpsharPensjonInput] = useState(
    convertBooleanToBooleanRadio(epsHarPensjon)
  )
  const [epsHarInntektOver2GInput, setEpsHarInntektOver2GInput] = useState(
    convertBooleanToBooleanRadio(epsHarInntektOver2G)
  )

  const formatertSivilstand = useMemo(
    () => formatSivilstand(intl, sivilstandFolkeregister).toLowerCase(),
    [sivilstandFolkeregister]
  )

  const shouldShowInput = React.useMemo(() => {
    const shouldShowEpsHarPensjon =
      sivilstandInput === 'GIFT' ||
      sivilstandInput === 'REGISTRERT_PARTNER' ||
      sivilstandInput === 'SAMBOER'

    const shouldShowEpsHarInntektOver2G =
      shouldShowEpsHarPensjon && epsHarPensjonInput === 'nei'

    if (!shouldShowEpsHarPensjon) {
      setValidationError({
        epsHarPensjon: undefined,
        epsHarInntektOver2G: undefined,
      })
    }

    return {
      epsHarPensjon: shouldShowEpsHarPensjon,
      epsHarInntektOver2G: shouldShowEpsHarInntektOver2G,
    }
  }, [sivilstandInput, epsHarPensjonInput])

  React.useEffect(() => {
    if (sivilstandInput !== null) {
      setValidationError((prev) => ({ ...prev, sivilstand: undefined }))
    }
    if (epsHarPensjonInput !== null) {
      setValidationError((prev) => ({ ...prev, epsHarPensjon: undefined }))
    }
    if (epsHarInntektOver2GInput !== null) {
      setValidationError((prev) => ({
        ...prev,
        epsHarInntektOver2G: undefined,
      }))
    }
  }, [sivilstandInput, epsHarPensjonInput, epsHarInntektOver2GInput])

  const onSubmit = (e: FormEvent<HTMLFormElement>): void => {
    e.preventDefault()
    const sivilstand_validationText = intl.formatMessage({
      id: 'stegvisning.sivilstand.select_validation_error',
    })
    const epsHarPensjon_validationText = intl.formatMessage(
      { id: 'stegvisning.sivilstand.epsHarPensjon.validation_error' },
      { sivilstand: getSivilstandTekst(intl, sivilstandInput) }
    )
    const epsHarInntektOver2G_validationText = intl.formatMessage(
      { id: 'stegvisning.sivilstand.epsHarInntektOver2G.validation_error' },
      { sivilstand: getSivilstandTekst(intl, sivilstandInput) }
    )
    if (
      checkUkjentSivilstand(sivilstandInput) ||
      (shouldShowInput.epsHarPensjon && epsHarPensjonInput === null) ||
      (shouldShowInput.epsHarInntektOver2G && epsHarInntektOver2GInput === null)
    ) {
      const validationErrorText = {
        sivilstand: checkUkjentSivilstand(sivilstandInput)
          ? sivilstand_validationText
          : undefined,
        epsHarPensjon:
          shouldShowInput.epsHarPensjon && epsHarPensjonInput === null
            ? epsHarPensjon_validationText
            : undefined,
        epsHarInntektOver2G:
          shouldShowInput.epsHarInntektOver2G &&
          epsHarInntektOver2GInput === null
            ? epsHarInntektOver2G_validationText
            : undefined,
      }
      setValidationError(validationErrorText)
      if (validationErrorText.epsHarPensjon) {
        logger('skjema validering feilet', {
          skjemanavn: STEGVISNING_FORM_NAMES.sivilstand,
          data: intl.formatMessage({
            id: 'stegvisning.sivilstand.radio_epsHarPensjon_label',
          }),
          tekst: validationErrorText.epsHarPensjon,
        })
      }
      if (validationErrorText.epsHarInntektOver2G) {
        logger('skjema validering feilet', {
          skjemanavn: STEGVISNING_FORM_NAMES.sivilstand,
          data: intl.formatMessage({
            id: 'stegvisning.sivilstand.radio_epsHarInntektOver2G_label',
          }),
          tekst: validationErrorText.epsHarInntektOver2G,
        })
      }
      return
    }

    logger('button klikk', {
      tekst: `Neste fra ${paths.sivilstand}`,
    })

    setValidationError({
      epsHarPensjon: undefined,
      epsHarInntektOver2G: undefined,
    })

    onNext({
      sivilstand: sivilstandInput,
      epsHarPensjon: shouldShowInput.epsHarPensjon
        ? convertBooleanRadioToBoolean(epsHarPensjonInput)
        : null,
      epsHarInntektOver2G: shouldShowInput.epsHarInntektOver2G
        ? convertBooleanRadioToBoolean(epsHarInntektOver2GInput)
        : null,
    })
  }

  return (
    <Card hasLargePadding hasMargin>
      <form onSubmit={onSubmit}>
        <Heading level="2" size="medium" spacing>
          <FormattedMessage id="stegvisning.sivilstand.title" />
        </Heading>
        <BodyLong size="large" className={styles.ingress}>
          {checkUkjentSivilstand(sivilstandFolkeregister) ? (
            <FormattedMessage id="stegvisning.sivilstand.ingress_ukjent" />
          ) : (
            <>
              <FormattedMessage id="stegvisning.sivilstand.ingress_1" />
              {formatertSivilstand}
              <FormattedMessage id="stegvisning.sivilstand.ingress_2" />
            </>
          )}
        </BodyLong>
        <VStack gap="6">
          <Select
            className={styles.selectSivilstand}
            name="sivilstand"
            value={
              checkUkjentSivilstand(sivilstandInput) ? '' : sivilstandInput
            }
            onChange={(e) =>
              setSivilstandInput(e.target.value as UtvidetSivilstand)
            }
            label={intl.formatMessage({
              id: `stegvisning.sivilstand.select_label`,
            })}
            description={
              checkUkjentSivilstand(sivilstand)
                ? intl.formatMessage({
                    id: `stegvisning.sivilstand.select_description_ukjent`,
                  })
                : intl.formatMessage({
                    id: `stegvisning.sivilstand.select_description`,
                  })
            }
            error={validationError.sivilstand}
          >
            {checkUkjentSivilstand(sivilstandInput) && (
              <option disabled selected value="">
                {' '}
              </option>
            )}
            {sivilstandOptions.map((option) => (
              <option key={option} value={option}>
                {intl.formatMessage({
                  id: `sivilstand.${option}`,
                })}
              </option>
            ))}
          </Select>
          {shouldShowInput.epsHarPensjon && (
            <RadioGroup
              legend={intl.formatMessage(
                { id: 'stegvisning.sivilstand.radio_epsHarPensjon_label' },
                { sivilstand: getSivilstandTekst(intl, sivilstandInput) }
              )}
              description={intl.formatMessage({
                id: 'stegvisning.sivilstand.radio_epsHarPensjon_description',
              })}
              name="epsHarPensjon"
              value={epsHarPensjonInput}
              onChange={(value) => setEpsharPensjonInput(value)}
              error={validationError.epsHarPensjon}
              role="radiogroup"
              aria-required="true"
            >
              <Radio value="ja">
                <FormattedMessage id="stegvisning.sivilstand.radio_ja" />
              </Radio>
              <Radio value="nei">
                <FormattedMessage id="stegvisning.sivilstand.radio_nei" />
              </Radio>
            </RadioGroup>
          )}
          {shouldShowInput.epsHarInntektOver2G && (
            <RadioGroup
              legend={intl.formatMessage(
                {
                  id: 'stegvisning.sivilstand.radio_epsHarInntektOver2G_label',
                },
                {
                  sivilstand: getSivilstandTekst(intl, sivilstandInput),
                  grunnbelop: grunnbelop
                    ? ` (${formatInntekt(grunnbelop * 2)} kr)`
                    : '',
                }
              )}
              description={intl.formatMessage({
                id: 'stegvisning.sivilstand.radio_epsHarInntektOver2G_description',
              })}
              value={epsHarInntektOver2GInput}
              onChange={(value) => setEpsHarInntektOver2GInput(value)}
              name="epsHarInntektOver2G"
              error={validationError.epsHarInntektOver2G}
              role="radiogroup"
              aria-required="true"
            >
              <Radio value="ja">
                <FormattedMessage id="stegvisning.sivilstand.radio_ja" />
              </Radio>
              <Radio value="nei">
                <FormattedMessage id="stegvisning.sivilstand.radio_nei" />
              </Radio>
            </RadioGroup>
          )}
        </VStack>
        <Button type="submit" className={styles.button}>
          <FormattedMessage id="stegvisning.neste" />
        </Button>
        <Button
          type="button"
          className={styles.button}
          variant="secondary"
          onClick={wrapLogger('button klikk', {
            tekst: `Tilbake fra ${paths.sivilstand}`,
          })(onPrevious)}
        >
          <FormattedMessage id="stegvisning.tilbake" />
        </Button>
        {onCancel && (
          <Button
            type="button"
            className={styles.button}
            variant="tertiary"
            onClick={wrapLogger('button klikk', { tekst: 'Avbryt' })(onCancel)}
          >
            <FormattedMessage id="stegvisning.avbryt" />
          </Button>
        )}
      </form>
    </Card>
  )
}
