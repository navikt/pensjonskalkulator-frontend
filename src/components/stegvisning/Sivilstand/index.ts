export { Sivilstand } from './Sivilstand'
export type { SivilstandRadio } from './Sivilstand'
