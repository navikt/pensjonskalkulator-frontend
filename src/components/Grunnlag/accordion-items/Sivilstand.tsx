import { useMemo } from 'react'

import { Accordion, Alert, BodyLong, Link } from '@navikt/ds-react'

import { formatSivilstand } from '@/components/Grunnlag/accordion-items/Sivilstand-utils'
import { useGetPersonQuery } from '@/state/api/apiSlice'

import { SectionHeader } from './components/SectionHeader'

export function Sivilstand() {
  const { data: person, isError, isLoading, isSuccess } = useGetPersonQuery()

  const formatertSivilstand = useMemo(
    () => (person ? formatSivilstand(person.sivilstand) : ''),
    [person]
  )

  if (isLoading) {
    return null
  }

  if (isError || !isSuccess) {
    return <Alert variant="error">Kunne ikke hente sivilstand</Alert>
  }

  return (
    <Accordion.Item data-testid={'accordion-sivilstand'}>
      <SectionHeader label="Sivilstand" value={formatertSivilstand} />
      <Accordion.Content>
        <BodyLong>
          Garantipensjon skal sikre et minste garantipensjonsnivå ved 67 år for
          de som har lav eller ingen opptjening til inntektspensjon. Størrelsen
          på garantipensjonen avhenger av din sivilstatus, og om ektefelle,
          partner eller samboer har egen inntekt eller mottar egen pensjon.{' '}
          <Link>Mer om garantipensjon og satser</Link>.
        </BodyLong>
      </Accordion.Content>
    </Accordion.Item>
  )
}
