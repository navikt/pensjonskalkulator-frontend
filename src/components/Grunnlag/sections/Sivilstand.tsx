import { useMemo } from 'react'

import { Accordion, BodyLong, Link } from '@navikt/ds-react'

import { useGetPersonQuery } from '@/state/api/apiSlice'

import { SectionContent } from './components/SectionContent'
import { SectionHeader } from './components/SectionHeader'
import { formatSivilstand } from './Sivilstand-utils'

export function Sivilstand() {
  const { data: person, isSuccess } = useGetPersonQuery()

  const formatertSivilstand = useMemo(
    () =>
      person && person.sivilstand ? formatSivilstand(person.sivilstand) : '',
    [person]
  )

  return (
    <Accordion.Item data-testid="accordion-sivilstand">
      <SectionHeader
        label="Sivilstand"
        value={
          isSuccess && person.sivilstand !== null
            ? formatertSivilstand
            : 'Kunne ikke hentes'
        }
      />
      <SectionContent>
        <BodyLong>
          Garantipensjon skal sikre et minste garantipensjonsnivå ved 67 år for
          de som har lav eller ingen opptjening til inntektspensjon. Størrelsen
          på garantipensjonen avhenger av din sivilstatus, og om ektefelle,
          partner eller samboer har egen inntekt eller mottar egen pensjon.{' '}
          <Link>Mer om garantipensjon og satser</Link>.
        </BodyLong>
      </SectionContent>
    </Accordion.Item>
  )
}
