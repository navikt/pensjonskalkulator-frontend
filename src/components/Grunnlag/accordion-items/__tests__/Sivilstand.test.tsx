import { Sivilstand } from '../Sivilstand'
import { mockErrorResponse, mockResponse } from '@/mocks/server'
import { render, screen, waitFor } from '@/test-utils'

describe('Sivilstand', () => {
  it('viser riktig tekst og lenke når henting av sivilstand er vellykket', async () => {
    mockResponse('/person', { status: 200, json: { sivilstand: 'GIFT' } })
    render(<Sivilstand />)

    await waitFor(() => {
      const el = screen.getByTestId('accordion-sivilstand')
      expect(el).toMatchSnapshot()
    })
  })

  test.each([
    ['GIFT', 'Gift'],
    ['SKILT', 'Skilt'],
    ['SEPARERT', 'Separert'],
    ['REGISTRERT_PARTNER', 'Registrert partner'],
    ['SEPARERT_PARTNER', 'Separert partner'],
    ['SKILT_PARTNER', 'Skilt partner'],
    ['GJENLEVENDE_PARTNER', 'Gjenlevende partner'],
    ['UGIFT', 'Ugift'],
  ])('viser riktig tekst når sivilstand er: %i', async (a, expected) => {
    mockResponse('/person', { status: 200, json: { sivilstand: a } })
    render(<Sivilstand />)
    await waitFor(() => {
      expect(screen.getByText(expected)).toBeVisible()
    })
  })

  it('viser feilmelding når henting av sivilstand feiler', async () => {
    mockErrorResponse('/person')
    render(<Sivilstand />)

    await waitFor(() => {
      expect(screen.getByText('Kunne ikke hente sivilstand')).toBeVisible()
    })
  })
})
