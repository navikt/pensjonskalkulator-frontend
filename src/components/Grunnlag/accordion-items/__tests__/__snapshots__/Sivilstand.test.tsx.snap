// Vitest Snapshot v1

exports[`Sivilstand > viser riktig tekst og lenke når henting av sivilstand er vellykket 1`] = `
<div
  class="navds-accordion__item"
  data-testid="accordion-sivilstand"
>
  <button
    aria-expanded="false"
    class="navds-accordion__header"
    type="button"
  >
    <div
      class="navds-accordion__icon-wrapper"
    >
      <svg
        aria-hidden="true"
        aria-labelledby="title-r0"
        class="navds-accordion__header-chevron"
        fill="none"
        focusable="false"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="title-r0"
        >
          Vis mer
        </title>
        <path
          clip-rule="evenodd"
          d="M5.97 9.47a.75.75 0 0 1 1.06 0L12 14.44l4.97-4.97a.75.75 0 1 1 1.06 1.06l-5.5 5.5a.75.75 0 0 1-1.06 0l-5.5-5.5a.75.75 0 0 1 0-1.06Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <span
      class="navds-accordion__header-content navds-heading navds-heading--small"
    >
      Sivilstand
      : 
      <span
        class="_details_8a9925"
      >
        Gift
      </span>
    </span>
  </button>
  <div
    class="navds-accordion__content navds-accordion__content--closed navds-body-long"
  >
    <p
      class="navds-body-long"
    >
      Garantipensjon skal sikre et minste garantipensjonsnivå ved 67 år for de som har lav eller ingen opptjening til inntektspensjon. Størrelsen på garantipensjonen avhenger av din sivilstatus, og om ektefelle, partner eller samboer har egen inntekt eller mottar egen pensjon.
       
      <a
        class="navds-link"
      >
        Mer om garantipensjon og satser
      </a>
      .
    </p>
  </div>
</div>
`;
